### PR TITLE
Pruning

### PR DESCRIPTION
**Pull request by Alex6323**
# Description of change

Adds pruning functionality.

## Links to any relevant issues

#497 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

VPS (mainnet and testnet)

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

File: bee-ledger/src/workers/consensus/worker.rs [hunk 1]
```diff
@@ -334,18 +338,30 @@ where
...
                         }
 
-                        if should_snapshot(&tangle, MilestoneIndex(*ledger_index), depth, &snapshot_config) {
-                            // TODO
-                            // if let Err(e) = snapshot(snapshot_config.path(), event.index - depth) {
-                            //     error!("Failed to create snapshot: {:?}.", e);
-                            // }
+                        match should_snapshot(&tangle, ledger_index, pruning_delay_min, &snapshot_config) {
+                            Ok(()) => {
+                                // TODO
```
Comment (samuel-rufi) [inline bee-ledger/src/workers/consensus/worker.rs:348]: Does this mean that the node won't create snapshots on his own? This PR just focuses on pruning right? Snapshoting will be handled in a further PR?

File: bee-ledger/src/workers/pruning/prune.rs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        pruning_metrics.indexations += unconfirmed_metrics.prunable_indexations;
+
+        // Remove old SEPs from the storage.
+        //
+        // **NOTE**: This operation must come before the batch is committed.
+        //
+        // TODO: consider batching deletes rather than using Truncate. Is one faster than the other? Do we care if its
+        // atomic or not?
+        let truncate_old_seps = Instant::now();
+        Truncate::<SolidEntryPoint, MilestoneIndex>::truncate(storage).unwrap();
```
Comment (samuel-rufi) [inline bee-ledger/src/workers/pruning/prune.rs]: uncommented/unhandled `unwrap()`

File: bee-ledger/src/workers/pruning/condition.rs [hunk 3]
```diff
@@ -1,77 +1,50 @@
-// Copyright 2020-2021 IOTA Stiftung
+// Copyright 2020 IOTA Stiftung
```
Comment (samuel-rufi) [inline bee-ledger/src/workers/pruning/condition.rs]: Seems like multiple Copyright headers are not consistent, please check the files.

**Review by samuel-rufi — CHANGES_REQUESTED** [review]

File: bee-ledger/src/workers/consensus/worker.rs (hunk 1 above)
Comment (thibault-martinez) [reply bee-ledger/src/workers/consensus/worker.rs:348]: Yes, just not a priority at all at the moment since we can rely on the uploaded snapshots.

Comment (samuel-rufi) [reply bee-ledger/src/workers/consensus/worker.rs:348]: okey 👍

File: bee-ledger/src/workers/pruning/error.rs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+use bee_message::{prelude::MilestoneIndex, MessageId};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Minimum: {minimum}, Found: {found}")]
+    InvalidTargetIndex {
+        minimum: MilestoneIndex,
+        found: MilestoneIndex,
+    },
+
```
Comment (pvdrz) [inline bee-ledger/src/workers/pruning/error.rs]: do we need these spaces?

**Review by pvdrz — DISMISSED** [review]

File: bee-ledger/src/workers/pruning/error.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use bee_message::{prelude::MilestoneIndex, MessageId};
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Minimum: {minimum}, Found: {found}")]
```
Comment (Adam-Gleave) [inline bee-ledger/src/workers/pruning/error.rs]: Everyone's favourite question -- what's the convention here, are the capitals fine to use? Same with the rest of the errors here.

**Review by Adam-Gleave — DISMISSED** [review]

File: bee-ledger/src/workers/consensus/worker.rs (hunk 1 above)
Comment (Alex6323) [reply bee-ledger/src/workers/consensus/worker.rs:348]: yea, it just focuses on pruning for now. I have a look at how the snapshot file is being dealt with. Maybe I can add this quickly to this PR.

File: bee-ledger/src/workers/pruning/error.rs (hunk 5 above)
Comment (pvdrz) [reply bee-ledger/src/workers/pruning/error.rs]: true, it should be no leading capitalization and no dot at the end.

Comment (Alex6323) [reply bee-ledger/src/workers/pruning/error.rs]: fixed

File: bee-ledger/src/workers/pruning/condition.rs (hunk 3 above)
Comment (Alex6323) [reply bee-ledger/src/workers/pruning/condition.rs]: @thibault-martinez what's the way/policy here?

File: bee-ledger/src/workers/pruning/error.rs (hunk 4 above)
Comment (Alex6323) [reply bee-ledger/src/workers/pruning/error.rs]: you mean the empty lines between variants?

File: bee-ledger/Cargo.toml [hunk 6]
```diff
@@ -22,9 +22,11 @@ async-trait = { version = "0.1", optional = true }
 chrono = { version = "0.4", optional = true }
 digest = { version = "0.9", optional = true }
 futures = { version = "0.3", optional = true }
+hashbrown = "0.11.2"
```
Comment (thibault-martinez) [inline bee-ledger/Cargo.toml]:
```suggestion
hashbrown = "0.11"
```

File: bee-ledger/Cargo.toml [hunk 7]
```diff
@@ -22,9 +22,11 @@ async-trait = { version = "0.1", optional = true }
...
 hex = { version = "0.4", optional = true }
 iota-crypto = { version = "0.6.0", features = [ "blake2b" ], optional = true }
 log = { version = "0.4", optional = true }
+ref-cast = "1.0.6"
```
Comment (thibault-martinez) [inline bee-ledger/Cargo.toml]:
```suggestion
ref-cast = "1.0"
```

File: bee-ledger/src/workers/pruning/error.rs (hunk 4 above)
Comment (thibault-martinez) [reply bee-ledger/src/workers/pruning/error.rs]: I would remove them, a bit inconsistent with the majority of the codebase

File: bee-message/src/message.rs [hunk 8]
```diff
@@ -68,6 +68,11 @@ impl Message {
     pub fn nonce(&self) -> u64 {
         self.nonce
     }
+
+    /// Destroys the `Message`, and returns ownership over its `Parents`.
```
Comment (thibault-martinez) [inline bee-message/src/message.rs]:
```suggestion
    /// Consumes the `Message`, and returns ownership over its `Parents`.
```

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: bee-ledger/src/workers/pruning/mod.rs [hunk 9]
```diff
@@ -1,192 +1,13 @@
-// Copyright 2020-2021 IOTA Stiftung
+// Copyright 2021 IOTA Stiftung
```
Comment (thibault-martinez) [inline bee-ledger/src/workers/pruning/mod.rs]: Keep 2020 if the file was already there

File: bee-ledger/src/workers/pruning/config.rs [hunk 10]
```diff
@@ -1,12 +1,17 @@
-// Copyright 2020-2021 IOTA Stiftung
```
Comment (thibault-martinez) [inline bee-ledger/src/workers/pruning/config.rs:1]: Keep 2020 if the file was already there

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**